src/util/result.h

```diff
@@ -8,16 +8,149 @@
 #include <attributes.h>
 #include <util/translation.h>
 
+#include <memory>
+#include <optional>
+#include <tuple>
+#include <utility>
 #include <variant>
+#include <vector>
 
 namespace util {
+namespace detail {
+//! Empty string list
+const std::vector<bilingual_str> EMPTY_LIST{};
 
+//! Helper function to join messages in space separated string.
+bilingual_str JoinMessages(const std::vector<bilingual_str>& errors, const std::vector<bilingual_str> warnings);
+
+//! Helper function to move messages from one vector to another.
+void MoveMessages(std::vector<bilingual_str>& src, std::vector<bilingual_str>& dest);
+
+//! Error information only allocated if there are errors or warnings.
+template <typename F>
+struct ErrorInfo {
+    std::optional<std::conditional_t<std::is_same<F, void>::value, std::monostate, F>> failure{};
+    std::vector<bilingual_str> errors;
+    std::vector<bilingual_str> warnings;
+};
+
+//! Result base class which is inherited by Result<T, F>.
+//! T is the type of the success return value, or void if there is none.
+//! F is the type of the failure return value, or void if there is none.
+template <typename T, typename F>
+class ResultBase;
+
+//! Result base specialization for empty (T=void) value type. Holds error
+//! information and provides accessor methods.
+template <typename F>
+class ResultBase<void, F>
+{
+protected:
+    std::unique_ptr<ErrorInfo<F>> m_info;
+
+    ErrorInfo<F>& Info()
+    {
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        return *m_info;
+    }
+
+    //! Value accessors that do nothing this because class is specializing value type T=void.
+    void ConstructValue() {}
+    template <typename O>
+    void MoveValue(O& other) {}
+    void DestroyValue() {}
+
+    template <typename, typename>
+    friend class ResultBase;
+
+public:
+    void AddError(bilingual_str error)
+    {
+        if (!error.empty()) Info().errors.emplace_back(std::move(error));
+    }
+
+    void AddWarning(bilingual_str warning)
+    {
+        if (!warning.empty()) Info().warnings.emplace_back(std::move(warning));
+    }
+
+    //! Operator moving warning and error messages from other result to this.
+    //! Only moves message strings, does not change success or failure values of
+    //! either Result object.
+    template<typename O>
+    O&& operator<<(O&& other)
+    {
+        if (other.m_info) {
+            if (!other.m_info->errors.empty()) MoveMessages(other.m_info->errors, Info().errors);
+            if (!other.m_info->warnings.empty()) MoveMessages(other.m_info->warnings, Info().warnings);
+        }
+        return std::forward<O>(other);
+    }
+
+    //! Success check.
+    explicit operator bool() const { return !m_info || !m_info->failure; }
+
+    //! Error retrieval.
+    template <typename _F = F>
+    std::enable_if_t<!std::is_same<_F, void>::value, const _F&> GetFailure() const { assert(!*this); return *m_info->failure; }
```

Comment:
nit: Can use `is_same_v` instead of `...::value`? Also, missing LIFETIMEBOUND?

(Same feedback on other lines where this is applicable)